File: composeApp/src/iosMain/kotlin/org/example/kmpsamples/presentation/permissions/systemPermissionHandlers/CalendarLegacyHandler.kt
package org.example.kmpsamples.presentation.permissions.systemPermissionHandlers

import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.useContents
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import org.example.kmpsamples.permissions.systemPermissionHandlers.SystemPermissionHandler
import org.example.kmpsamples.presentation.permissions.dtos.Permission
import org.example.kmpsamples.presentation.permissions.dtos.Permission.CalendarReadPermission
import org.example.kmpsamples.presentation.permissions.dtos.Permission.CalendarWritePermission
import org.example.kmpsamples.presentation.permissions.dtos.PermissionState
import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType
import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType.MultiplePermissionState
import org.example.kmpsamples.presentation.permissions.dtos.PermissionStateType.SinglePermissionState
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType.PlatformSinglePermission
import platform.EventKit.EKAuthorizationStatusAuthorized
import platform.EventKit.EKAuthorizationStatusDenied
import platform.EventKit.EKAuthorizationStatusNotDetermined
import platform.EventKit.EKAuthorizationStatusRestricted
import platform.EventKit.EKEntityType
import platform.EventKit.EKEventStore
import platform.Foundation.NSProcessInfo

class CalendarLegacyHandler() : SystemPermissionHandler {
    private val mainScope = MainScope()
    private val eventStore = EKEventStore()

    override fun canHandle(simplePlatformPermission: PlatformSinglePermission): Boolean {
        fun isSupported(permission: Permission): Boolean = when (permission) {
            is CalendarReadPermission, is CalendarWritePermission -> !isAtLeastIOS(
                17
            )

            else -> false
        }
        return isSupported(simplePlatformPermission.permission)
    }

    private fun createMultiplePermissionState(
        isGranted: Boolean,
        shouldShowRationale: Boolean
    ): MultiplePermissionState {
        val permissionStates = listOf(CalendarReadPermission(), CalendarWritePermission()).map {
            val permissionState = PermissionState(it, isGranted, shouldShowRationale)
            SinglePermissionState(permissionState)
        }
        return MultiplePermissionState(permissionStates)
    }

    override fun resolvePermissionState(): PermissionStateType {
        val status = EKEventStore.authorizationStatusForEntityType(EKEntityType.EKEntityTypeEvent)
        return when (status) {
            EKAuthorizationStatusAuthorized -> createMultiplePermissionState(true, false)
            EKAuthorizationStatusDenied -> createMultiplePermissionState(false, false)
            EKAuthorizationStatusNotDetermined -> createMultiplePermissionState(false, true)
            EKAuthorizationStatusRestricted -> createMultiplePermissionState(false, false)
            else -> createMultiplePermissionState(false, true)
        }
    }

    @OptIn(ExperimentalForeignApi::class)
    private fun isAtLeastIOS(version: Int): Boolean {
        val majorVersion =
            NSProcessInfo.processInfo.operatingSystemVersion.useContents { majorVersion }
        return majorVersion >= version
    }

    override fun requestPermission(onRequest: (permissionState: PermissionStateType) -> Unit) {
        eventStore.requestAccessToEntityType(EKEntityType.EKEntityTypeEvent) { isGranted, error ->
            mainScope.launch {
                onRequest(resolvePermissionState())
            }
        }
    }
}

File: iosApp/iosApp/Permissions/SystemPermissionHandlers/CalendarLegacyHandler.swift
import EventKit
import Foundation
import ComposeApp

class CalendarLegacyHandler: SystemPermissionHandler {
    var id: String = "CalendarLegacyHandler"
    
    private let eventStore = EKEventStore()
    
    func canHandle(simplePlatformPermission: PlatformPermissionTypePlatformSinglePermission) -> Bool {
        func isSupported(permission: Permission) -> Bool {
            switch permission {
            case is Permission.CalendarReadPermission, is Permission.CalendarWritePermission:
                return !isAtLeastIOS(version: 17)
            default:
                return false
            }
        }
        return isSupported(permission: simplePlatformPermission.permission)
    }
    
    private func createMultiplePermissionState(isGranted: Bool, shouldShowRationale: Bool) -> PermissionStateTypeMultiplePermissionState {
        let permissionStates = [Permission.CalendarReadPermission(), Permission.CalendarWritePermission()].map {
            let permissionState = PermissionState(permission: $0, isGranted: isGranted, shouldShowRationale: shouldShowRationale)
            return PermissionStateTypeSinglePermissionState(permissionState: permissionState)
        }
        return PermissionStateTypeMultiplePermissionState(permissionStates: permissionStates)
    }
    
    func resolvePermissionState() -> PermissionStateType {
        let status = EKEventStore.authorizationStatus(for: .event)
        switch status {
        case .authorized, .writeOnly, .fullAccess:
            return createMultiplePermissionState(isGranted: true, shouldShowRationale: false)
        case .denied, .restricted:
            return createMultiplePermissionState(isGranted: false, shouldShowRationale: false)
        case .notDetermined:
            return createMultiplePermissionState(isGranted: false, shouldShowRationale: true)
        @unknown default:
            return createMultiplePermissionState(isGranted: false, shouldShowRationale: true)
        }
    }
    
    private func isAtLeastIOS(version: Int) -> Bool {
        let majorVersion = ProcessInfo.processInfo.operatingSystemVersion.majorVersion
        return majorVersion >= version
    }
    
    func requestPermission(onRequest: @escaping (PermissionStateType) -> Void) {
        eventStore.requestAccess(to: .event) { isGranted, error in
            DispatchQueue.main.async {
                onRequest(self.resolvePermissionState())
            }
        }
    }
}
